### PR TITLE
Add memory and CPU metric

### PR DESCRIPTION
**Pull request by eranra**
This PR adds CPU and memory metric to measures
(also adding generic Measure function that allows to call any of the metric functions generically)

### CODE REVIEW

File: internal/metrics/client.go [hunk 1]
```diff
@@ -69,6 +77,33 @@ func NewClient(url string, timeout time.Duration) (Client, error) {
...
+}
+
+func (c *client) ProcessMEM(job string, duration model.Duration) (float64, error) {
+	query := fmt.Sprintf(
+		`sum(avg_over_time(process_resident_memory_bytes{job="%s"}[%s]) / 1000000)`, // in Mi (i.e. in Megabytes)
+		job, duration,
+	)
+
+	return c.executeScalarQuery(query)
+}
```
Comment (periklis) [inline internal/metrics/client.go:105]: The selected metric `process_virtual_memory_bytes` represents a golang process from a general Linux view, but it is not always useful to track in raw isolation. In detail Golang's runtime is until 1.16 using `MADV_FREE` (See [blog post](https://www.bwplotka.dev/2019/golang-memory-monitoring/)) which translates in heavy virtual memory usage w/o giving back to the OS except on memory pressure. This is a known pitfall monitoring golang programs thus rather than using the process collector's metrics, the best we can do is to track memory usage via cAdvisor. Latter is is an optional utility on running clusters.

In summary, it will be an extra user task to require cAdvisor on running clusters to benchmark Loki. IMHO if we want to track proper memory usage we should use `container_memory_working_set_bytes` if available. In addition we should give users a hint to install cadvisor if they want to track memory usage.

__note:__ Loki is still on golang 1.15 thus `MADV_FREE` is still applied.

File: internal/metrics/client.go [hunk 2]
```diff
@@ -69,6 +77,33 @@ func NewClient(url string, timeout time.Duration) (Client, error) {
...
+
+func (c *client) ProcessCPU(job string, duration model.Duration) (float64, error) {
+	query := fmt.Sprintf(
+		`sum(rate(process_cpu_seconds_total{job="%s"}[%s]) * 1000)`, // in m (i.e. 1000 = 1 vCore)
+		job, duration,
+	)
+
+	return c.executeScalarQuery(query)
+}
```
Comment (periklis) [inline internal/metrics/client.go:96]: As mentioned below, if we go with optionally requiring cAdvisor we should use `container_cpu_user_seconds_total` instread of `process_cpu_seconds_total`

File: internal/metrics/client.go (hunk 1 above)
Comment (eranra) [reply internal/metrics/client.go:105]: @periklis between the sec you wrote the text and the changes I did to merge, the code has changes ... if you look on your remark you will see that the metric I use is process_resident_memory_bytes. This metric tracks the amount of memory used by the process ... I compared that with `oc adm top pod` values and saw that I am getting the same numbers ... 

Also, I do not see `container_memory_working_set_bytes`  exposed ... so we can't use that

File: internal/metrics/client.go (hunk 2 above)
Comment (eranra) [reply internal/metrics/client.go:96]: I do not see that metric exposed by the process (same as bellow, I compared with `oc adm top pod` and see same numbers so this metric is good for what we are trying to measure)

File: internal/metrics/client.go (hunk 1 above)
Comment (eranra) [reply internal/metrics/client.go:105]: I do not disagree with you that it is hard to track memory with golang aps ... but I think that for ower current needs since we are slowly increasing the stress and the need for memory the current implementation is sufficient. 

What typically happens is that the use case requires to redeploy Loki with different configuration thus recreate the relevant microservices. With that ... you zero memory and CPU and start all over again so the results we get are similar to what is really consumed

Comment (periklis) [reply internal/metrics/client.go:105]: As written above `container_memory_working_set_bytes` is only available on clusters with [cAdvisor](https://github.com/google/cadvisor). cAdvisor is a widespread addon on k8s clusters, because it gives you a sold view of your containers.

Resident memory is not the same on what you see on `oc adm top pod`. Latter is using `PodMetrics` from the k8s metrics-server which in turn is scrapping the container runtime for CPU/Memory metrics. For example on OCP we use `cri-o` as the default container runtime and in turn the memory metric passed to the metrics server is [working_set_bytes](https://github.com/cri-o/cri-o/blob/v1.21.0/server/container_stats.go#L46) which is the same as cAdvisor exposes as `container_memory_working_set_bytes`.

In general if we want these metrics to be part of `loki-benchmarks` we should be more opinionated on which metrics to pick. Me is hesitant to use the process collector because it gives a biased view. It is a considerable requirement for the user of this small bench tool to provide cAdvisor, but if not the graphs will/can remain silent for him/her.

Comment (eranra) [reply internal/metrics/client.go:105]: @periklis To me this feels like (1) overkill (2) adding more and more dependencies to this benchmark is not such a great idea ... I do not like the port forwarding and that "game" with Promethous and now with cAdvisor it makes things even more complicated. 

If you want I can change the name of the current metric from ProcessMEM to ProcessResidentMemory ... but I am not planning to add cAdvisor as part of this PR ... we can add another PR for that when we are done with the current changes ... we still have another big PR for the multi-configuration to work on.

Comment (periklis) [reply internal/metrics/client.go:105]: Pardon sir, I may not have expressed my thoughts clearly enough. I do not suggest to add cAdvisor as a hard dependency to `loki-benchmarks`. I suggest only to collect the cAdvisor `container_memory_working_set_bytes` instead of RSS as part of the measure call. The user looking on the bench results will see values/graphs if the target cluster includes cAdvisor. We can state this on the `README.md` or else the measurements will be zero.

IMHO adding a metric for RSS that is biased is not a helpful addition to `loki-benchmarks`. It is simply put not helpful enough to interpret Loki's memory usage for any benchmark scenario. I argue to not add collection of these metrics if we don't want to put a requirement to the users provide cAdvisor's `container_memory_working_set_bytes`.

Comment (eranra) [reply internal/metrics/client.go:105]: @periklis `... If you are limiting the resource usage for your pods than you should monitor both as they will cause an oom-kill if they reach a particular resource limit. ...` 

Read here: https://stackoverflow.com/questions/65428558/what-is-the-difference-between-container-memory-working-set-bytes-and-contain

I do not think that this is misleading if we put the correct label  / tag on the data 

Let me refine my proposal ... we will call the metric --> "ProcessMemoryResidentSetSize " 

and in following later PR add text that explains how to deploy cAdvisor and add the code to monitor also that matrics ... does this make sense ???

Comment (periklis) [reply internal/metrics/client.go:105]: Taking the article into account, it also states:
```
The better metric is container_memory_working_set_bytes as this is what the OOM killer is watching for.
```
Thus looking on RSS only doesn't give you a clear hint on when the OOM kill happens. RSS can tell you that you hit the limit including all the inactive but not yet evicted pages, but the OOM Killer will not care.

File: internal/metrics/client.go (hunk 2 above)
Comment (periklis) [reply internal/metrics/client.go:97]: As per adding cavidsor, we can use `container_cpu_user_seconds_total` now, WDYT?

File: internal/metrics/client.go [hunk 3]
```diff
@@ -69,6 +78,45 @@ func NewClient(url string, timeout time.Duration) (Client, error) {
...
+// NOTE: Using ProcessResidentMEM is not recommended (information not representative for golang apps)
+// It is recommended to deploy cadvisor and use ProcessWorkingSetMEM
+func (c *client) ProcessResidentMEM(job string, duration model.Duration) (float64, error) {
+	query := fmt.Sprintf(
+		`sum(avg_over_time(process_resident_memory_bytes{job="%s"}[%s]) / 1000000)`, // in Mi (i.e. in Megabytes)
+		job, duration,
+	)
+
+	return c.executeScalarQuery(query)
+}
```
Comment (periklis) [inline internal/metrics/client.go:118]: Does this imply a fallback metric for target systems under test which don't provide cadvisor metrics? If not, should this be removed?

**Review by periklis — COMMENTED** [review]
Do you intend to complete scraping CPU/MEM metrics for all components or keep it only for ingesters? If latter, why not?

File: internal/metrics/client.go (hunk 3 above)
Comment (eranra) [reply internal/metrics/client.go]: @periklis Yes ... in the future I think we should add this as a fallback (also the same thing for memory)

**Comment by eranra** [comment]
> Do you intend to complete scraping CPU/MEM metrics for all components or keep it only for ingesters? If latter, why not?

For now, just for ingesters ... the problem is that we have "too much info" in the results folder ... maybe it makes sense to organize the results in sub-folders and this will make things more understandable

**Comment by eranra** [comment]
> As per adding cavidsor, we can use container_cpu_user_seconds_total now, WDYT?

Added function with "that (renamed the functions to "container_YYY_" )

**Review by periklis — DISMISSED** [review]
LGTM

**Review by periklis — APPROVED** [review]

**Review by periklis — APPROVED** [review]
